test/eth.subscribe.ganache.js

```diff
@@ -35,6 +35,61 @@ describe('subscription connect/reconnect', function() {
         });
     });
 
+    it('subscribes with a callback', function(){
+        let subscription;
+
+        return new Promise(async function (resolve) {
+            subscription = web3.eth
+                .subscribe('newBlockHeaders', function(err, result){
+                    assert(result.parentHash);
+                    subscription.unsubscribe(); // Stop listening..
+                    resolve();
+            });
+         })
+    });
+
+    it('resubscribes', function(){
+        let subscription;
+
+        return new Promise(async function (resolve) {
+            subscription = web3.eth.subscribe('newBlockHeaders');
+            subscription.unsubscribe();
+            subscription.resubscribe()
+
+            subscription.on('data', function(result){
+                assert(result.parentHash);
+                subscription.unsubscribe(); // Stop listening..
+                resolve();
+            })
+         })
+    });
+
+    it('allows a subscription which does not exist', function(){
+        web3.eth.subscribe('subscription-does-not-exists');
+    });
+
+    it('errors when zero params subscrip. is called with the wrong arguments', function(){
```

Comment:
Is it possible to test here the ``error`` event listener and the callback ``error`` parameter?

Reply:
Apparently not - neither hears anything. An error is thrown.

Reply:
The passed callback should get called successfully with the error. In the case of the error event listener do we have a timing issue we already noticed when the ``provider.send`` call is failing before the ``EventEmitter`` got returned. We would have to add a synthetic delay or another solution to fix this. 🤔

Reply:
The same problem should exist in this case as well: https://github.com/ethereum/web3.js/blob/provider-related-improvements/packages/web3-core-subscriptions/src/subscription.js#L198

Reply:
@nivida What do you think about instead of firing the callbacks / emitters just throwing in those cases. 

It seems more reliable and it's before anything is sent - maybe it makes sense to "fail hard and fast" everywhere within [this code block](https://github.com/ethereum/web3.js/blob/ac880f656b564bf275ff2d417c3f2d62830a9ea2/packages/web3-core-subscriptions/src/subscription.js#L191-L218) at `.subscribe` .

Reply:
Yes, this would probably be the cleanest solution but would introduce a breaking-change because all DApp devs would have to add a try/catch around the ``subscribe`` call and they would get errors where they were silently ignored before. 🤔

Reply:
Additionally, would this mean that some errors are thrown, some are only returned in the callback, and some are emitted through with the error event and are returned to the callback.

Reply:
Ah yes! 

In d21d746d2567f0cf7adc6bf699404d9bae3bf11b have wrapped the 'no provider' and 'no provider.on' handling in a zero-length timeout (to schedule the emitter in the next execution loop) and added tests for all those cases. 

Preserved the current handling for bad params because that's the existing behavior. Not sure there's much value in adding two other ways that error can express itself - we'd still need to throw if people currently expect that. Also seems consistent with much of the other param validation.  

What is your view of it?

Reply:
The currently applied fix does look good to me 👍 

Is it possible to fix the error handling on line 221 and 247 in the ``subscription.js`` file the same way as it's done for the parameter validation in this PR as well? 

Because when the call of ``requestManager.send`` does fail immediately do we have the same problem as we have for the parameter validation. This is probably an edge-case but it will definitely happen in some dev/test setups.

Reply:
Added in 9038a7a9d4a6872234302d933bb6517a26052c2b.